parlai/core/logging_utils.py

```diff
@@ -0,0 +1,132 @@
+        :param name: Name of the logger
+        :param console_level: min. Level of messages logged to console
+        :param console_format: The format of messages logged to the console. Simple stdout is used if None specified
+        :param file_format: The format of messages logged to the file (A default is used if None specified)
+        :param file_level: min. Level of messages logged to the file
+        :param filename: The file the logs are written to
+        """
+        super().__init__(name, console_level)  # can be initialized with any level
+
+        # Logging to a file
+        if filename:
+            # Default format used if no file format provided
+            if file_format is None:
+                file_format = '%(asctime)s :: %(levelname)s :: %(message)s'
+            self.fileHandler = logging.FileHandler(filename)
+            self.fileHandler.level = (
+                file_level
+            )  # Log to file levels: file_level and above
+            self.fileFormatter = logging.Formatter(file_format)
+            self.fileHandler.setFormatter(self.fileFormatter)
+            super().addHandler(self.fileHandler)
+
+        # Logging to stdout
+        self.streamHandler = logging.StreamHandler(sys.stdout)
+        self.streamHandler.level = (
+            console_level
+        )  # Log to stdout levels: console_level and above
+        if console_format is not None:
+            self.consoleFormatter = logging.Formatter(console_format)
+            self.streamHandler.setFormatter(self.consoleFormatter)
+        super().addHandler(self.streamHandler)
+
+        # To be used with testing_utils.capture_output()
+        self.altStream = None
+
+    def log(self, msg, level=INFO):
+        """ Default Logging function """
+        super().log(level, msg)
+
+    def add_file_handler(self, filename, level=INFO, format=None):
+        """
+        Add a file handler to the logger object
+        Use case: When logging using the logger object instead of instantiating a new ParlaiLogger
+                  this function might  be useful to add a filehandler on the go
+        Only does so if there is no file handler existing
+        """
+        if not hasattr(self, 'fileHandler'):
+            if format is None:
+                file_format = '%(asctime)s :: %(levelname)s :: %(message)s'
+            self.fileHandler = logging.FileHandler(filename)
+            self.fileHandler.level = level  # Log to file levels: level and above
+            self.fileFormatter = logging.Formatter(file_format)
+            self.fileHandler.setFormatter(self.fileFormatter)
+            super().addHandler(self.fileHandler)
+        else:
+            self.info("ParlaiLogger: A filehandler already exists")
+
+    def add_format_prefix(self, prefix):
+        """Include `prefix` in all future logging statements"""
+        # change both handler formatters to add a prefix
+        new_str = prefix + " " + '%(message)s'
```

Comment:
is the assumption that the console logger only has '%(message)s' ever as its format? Maybe we can at least document that limitation at the top of the class.

Reply:
That's a good catch. Fixed it in my next commit to remove that assumption. Now the only assumption is that the format is divided by ```:``` because that's what I split on. Mentioning that on top of the class.